### PR TITLE
[quant][graphmode][fx] Add support for one value being quantized with different qconfigs

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54928 [quant][graphmode][fx][refactor] Quantize by Use of a Tensor instead of Tensor
* #54924 [quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat
* #54813 [quant][graphmode][fx] Optimize cat
* #54812 [nn] Add allow_duplicate option for named_modules
* #54733 [quant][graphmode][fx][refactor] Factor out insert_observers_for_model to a separate function
* #54644 [quant][graphmode][fx] Separate handling Copy operator to a helper function
* #54643 [quant][fx][graphmode][refactor] Change activation_post_process_map to track the observer name instead
* #54073 [quant][graphmode][refactor] Remove reduandent code
* **#53586 [quant][graphmode][fx] Add support for one value being quantized with different qconfigs**

Summary:
Previously one value can only be quantized to one dtype, this PR adds the support for quantizing one value
in the fx graph with multiple dtypes, e.g. first quantize to int8 and then float16

might do some followup PRs to clean up the hacks and refactor the code.

Test Plan:
python test/test_quantization.py TestQuantizeFx.test_multiple_qconfigs_single_value

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D26912676](https://our.internmc.facebook.com/intern/diff/D26912676)

### CODE REVIEW

File: torch/quantization/fx/quantize.py [hunk 1]
```diff
@@ -78,7 +80,7 @@
 def insert_observer(
         node: Node, observer: torch.quantization.ObserverBase,
         model: torch.nn.Module,
-        activation_post_process_map: Dict[str, torch.quantization.ObserverBase],
+        activation_post_process_map: Dict[str, List[torch.quantization.ObserverBase]],
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:83]: does the graph look like

```
          --> obs2
         /
node --> obs1
```

or 

```
node -> obs1 -> obs2
```

?

Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:83]: the second one

File: torch/quantization/fx/quantization_patterns.py [hunk 2]
```diff
@@ -181,7 +182,7 @@ def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
                     'call_function', self.quantized_binary_op,
                     load_arg(quantized=[quantized_index])(self.binary_op_node.args), self.binary_op_node.kwargs)
             else:
-                activation_post_process = quantizer.activation_post_process_map[node.name]
+                activation_post_process = quantizer.activation_post_process_map[node.name].pop(0)
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:185]: yes, this is a bit hacky, I plan to clean up in a later PR

[the remark this answers is not in the document]
Comment (vkuzo) [reply torch/quantization/fx/quantization_patterns.py:185]: makes sense, any chance we could make it happen in this PR, or move the refactor to be before this PR?  This seems like a pretty risky change

File: torch/quantization/fx/quantize.py [hunk 3]
```diff
@@ -1096,10 +1112,10 @@ def _find_quants(self, graph: Graph, matches: Dict[str, MatchResult],
           - matches: output of self._find_matches function
 
         Outputs a map of
-         node_name -> (QuantizeHandler instance (always DefaultQuantizeHandler),
+         node_name -> list of (QuantizeHandler instance (always DefaultQuantizeHandler),
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:1115]: can we make sure that the docblock explains why this is a list

File: torch/quantization/fx/quantization_patterns.py [hunk 4]
```diff
@@ -927,7 +929,12 @@ class CopyNode(QuantizeHandler):
     def convert(self, quantizer: QuantizerCls, node: Node, load_arg: Callable,
                 is_reference: bool = False,
                 convert_custom_config_dict: Dict[str, Any] = None) -> Node:
-        return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
+        qconfig = quantizer.qconfig_map[node.name]
+        if activation_is_int8_quantized(qconfig):
+            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=None))
+        else:
+            # fp16 static quantization
+            return quantizer.quantized_graph.node_copy(node, load_arg(quantized=False))
```
Comment (vkuzo) [inline torch/quantization/fx/quantization_patterns.py:937]: wondering on what's a good way to explain this?  Does `quantized` from `load_arg` mean the same thing as `activation_is_int8_quantized`?  If not, should we rename the `load_arg` argument to be more clear?

File: torch/quantization/fx/quantize.py [hunk 5]
```diff
@@ -669,12 +674,13 @@ def _convert(self, model: GraphModule, is_reference: bool = False,
...
 
-        quants: Dict[str, Tuple[DefaultQuantizeHandler, Callable]] = \
+        quants: Dict[str, List[Tuple[DefaultQuantizeHandler, Callable]]] = \
             self._find_quants(model.graph, matches)
 
         self.quantized_graph = Graph()
         env: Dict[str, Node] = {}
-        quant_env: Dict[str, Node] = {}
+        # TODO: merge quant_env with env
+        quant_env: Dict[str, Tuple[Node, torch.dtype]] = {}
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:683]: can we keep the types of `env` and `quant_env` consistent?  If `quant_env` needs a dtype, can we add the `torch.float` dtype to every entry of `env?

File: torch/quantization/fx/quantization_patterns.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:185]: I can't think of a easy way currently without making larger changes. I think it should be fine as long as all the tests passes. but please let me know if you have any suggestions.

File: torch/quantization/fx/quantization_patterns.py (hunk 4 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantization_patterns.py:937]: yeah this is a bit confusing, I'll think about how to make this more clear

File: torch/quantization/fx/quantize.py (hunk 5 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:683]: yeah, see TODO comment above, I plan to merge them later

File: torch/quantization/fx/quantize.py (hunk 3 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:1115]: sure

File: torch/quantization/fx/quantization_patterns.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (vkuzo) [reply torch/quantization/fx/quantization_patterns.py:185]: hmm, can we just do something like

```
cur_obs_idx = 0
... when needed, increment cur_obs_idx
cur_obs = quantizer.activation_post_process_map[node.name][cur_obs_idx]
```

it's a bit of extra bookkeeping, but it removes state mutations from this variable and instead moves them into the code

File: test/quantization/test_quantize_fx.py [hunk 6]
```diff
@@ -634,6 +634,7 @@ def forward(self, x):
         self.assertEqual(model_device, device)
 
     @skipIfNoFBGEMM
+    @unittest.skip("dict doesnot work in fx?")
```
Comment (vkuzo) [inline test/quantization/test_quantize_fx.py:637]: is it related to this PR?

File: torch/quantization/fx/quantize.py [hunk 7]
```diff
@@ -81,15 +83,19 @@
...
+        observed_node_names_set: Set[str],
+        quants: Dict[str, List[Tuple[DefaultQuantizeHandler, Callable]]]):
     """Insert observer for node by modifying the observed_graph and
        attach observer module to the model
        Args:
          node: Node
          observer: observer/fake_quantize module instance
     """
+    if not model.training and isinstance(observer, torch.quantization.FixedQParamsFakeQuantize):
+        return
```
Comment (vkuzo) [inline torch/quantization/fx/quantize.py:98]: just curious, what's the context on this one?  Maybe we can add a comment?

**Review by vkuzo — APPROVED** [review]
makes sense, accepting to unblock, would be great to undo the additional state in the future.  One idea I had when reading this PR...thoughts about this?
```
# map from (node.name, qconfig) to observer
self.activation_post_process_map: Dict[Tuple(str, QConfig), ObserverBase]
```
If the only use case for multiple observers for a node is multiple qconfigs, perhaps that can be used to deduplicate, and we wouldn't need to add an additional indices variable?  I haven't thought through it deeply, more asking for your early opinion.

File: test/quantization/test_quantize_fx.py (hunk 6 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_fx.py:637]: this test fails after the change, will need to ask FX team to help debug the issue

File: torch/quantization/fx/quantize.py (hunk 7 above)
Comment (jerryzh168) [reply torch/quantization/fx/quantize.py:98]: Sure I can add a comment. reason is fixedqparams fake quant only needs to be there for training, we can either do this check here or in _find_quant.

**Pull request closed without merging**